File: LeetCode/0992-Subarrays-with-K-Different-Integers.cpp
// 992. Subarrays with K Different Integers

class Solution {
public:

    int atMostK(vector<int>& nums, int k) {
        int n = nums.size(), l = 0, ans = 0;
        map<int,int> mp;
        for(int r = 0; r < n; r++) { 
            mp[nums[r]]++;
            while(l <= r && mp.size() > k) {
                mp[nums[l]]--;
                if(mp[nums[l]] == 0) mp.erase(nums[l]);
                l++;
            }
            ans += r - l + 1;
        }
        return ans;
    }

    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return atMostK(nums, k) - atMostK(nums, k - 1);
    }
};

File: LeetCode/0992-Subarrays-with-K-Different-Integers.java
// 992. Subarrays with K Different Integers

class Solution {
    public int subarraysUptoKDistinct(int[] nums, int k) {
        int n = nums.length;
        int l = 0, r = 0, ans = 0;
        Map<Integer,Integer> mp = new HashMap<>();
        while (r < n) {
            mp.put(nums[r], mp.getOrDefault(nums[r], 0) + 1);
            while (mp.size() > k) {
                mp.put(nums[l], mp.get(nums[l]) - 1);
                if (mp.get(nums[l]) == 0) mp.remove(nums[l]);
                l++;
            }
            ans += r - l + 1;
            r++;
        }
        return ans;
    }

    public int subarraysWithKDistinct(int[] nums, int k) {
        return subarraysUptoKDistinct(nums, k) - subarraysUptoKDistinct(nums, k - 1);
    }
}
